storage/gcloud/aio/storage/storage.py

```diff
@@ -612,17 +612,26 @@ async def _do_upload(self, session_uri: str, stream: IO[AnyStr],
                          timeout: int = 30) -> Dict[str, Any]:
         s = AioSession(session) if session else self.session
 
-        for tries in range(retries):
-            try:
-                resp = await s.put(session_uri, headers=headers,
-                                   data=stream, timeout=timeout)
-            except ResponseError:
-                headers.update({'Content-Range': '*/*'})
-                await sleep(2. ** tries)  # type: ignore[func-returns-value]
-
-                continue
-
-            break
+        original_close = stream.close
+        original_position = stream.tell()
+        # Prevent the stream being closed if put operation fails
+        stream.close = lambda: None  # type: ignore[assignment]
+        try:
+            for tries in range(retries):
+                try:
+                    resp = await s.put(session_uri, headers=headers,
+                                       data=stream, timeout=timeout)
+                except ResponseError:
+                    headers.update({'Content-Range': '*/*'})
+                    stream.seek(original_position)
+
+                    await sleep(  # type: ignore[func-returns-value]
+                        2. ** tries
```

Comment:
We're at 32s by the fifth try. This kinda bothers me because downstream users might end up with higher latency than their own apps are configured to tolerate, and have no recourse after they find out why they're timing out — we could potentially make the base `2.` an arg — but seeing as this is inherited, we can probably live with it.